lib/util/stdrandom.h

```diff
@@ -0,0 +1,369 @@
+{
+
+using std::numeric_limits;
+using std::true_type;
+using std::false_type;
+using std::conditional;
+using std::iterator_traits;
+using std::integral_constant;
+using std::swap;
+
+// This file contains extracts from libc++ <algoritm>, <random> and <bits> that
+// provide randomness-related components of std:: (specifically
+// uniform_int_distribution and shuffle) because these parts of the standard are
+// underspecified and vary between stdlibs. We therefore commit specifically to
+// some libc++ classes rather than using the stdlib-varying versions.
+
+// __libcpp_clz
+
+#if defined(_MSC_VER)
+#define _LIBCPP_COMPILER_MSVC
+#endif
+
+#if (defined(_M_AMD64) || defined(__x86_64__)) || (defined(_M_ARM) || defined(__arm__))
+#define _LIBCPP_HAS_BITSCAN64
+#endif
+
+#if defined(_LIBCPP_COMPILER_MSVC)
+#include <intrin.h>
+#endif
+
+#ifndef _LIBCPP_COMPILER_MSVC
+
+inline constexpr int
+__libcpp_clz(unsigned __x) noexcept
+{
+    return __builtin_clz(__x);
+}
+
+inline constexpr int
+__libcpp_clz(unsigned long __x) noexcept
+{
+    return __builtin_clzl(__x);
+}
+
+inline constexpr int
+__libcpp_clz(unsigned long long __x) noexcept
+{
+    return __builtin_clzll(__x);
+}
+
+#else  // _LIBCPP_COMPILER_MSVC
+// Precondition:  __x != 0
+inline
+int __libcpp_clz(unsigned __x) {
+  static_assert(sizeof(unsigned) == sizeof(unsigned long), "");
+  static_assert(sizeof(unsigned long) == 4, "");
+  unsigned long __where;
+  if (_BitScanReverse(&__where, __x))
+    return static_cast<int>(31 - __where);
+  return 32; // Undefined Behavior.
+}
+
+inline
+int __libcpp_clz(unsigned long __x) {
+    static_assert(sizeof(unsigned) == sizeof(unsigned long), "");
+    return __libcpp_clz(static_cast<unsigned>(__x));
+}
+
+inline
+int __libcpp_clz(unsigned long long __x) {
+  unsigned long __where;
+#if defined(_LIBCPP_HAS_BITSCAN64)
+  if (_BitScanReverse64(&__where, __x))
+    return static_cast<int>(63 - __where);
+#else
+  // Win32 doesn't have _BitScanReverse64 so emulate it with two 32 bit calls.
+  if (_BitScanReverse(&__where, static_cast<unsigned long>(__x >> 32)))
+    return static_cast<int>(63 - (__where + 32));
+  if (_BitScanReverse(&__where, static_cast<unsigned long>(__x)))
+    return static_cast<int>(63 - __where);
+#endif
+  return 64; // Undefined Behavior.
+}
+
+#endif
+
+
+// __independent_bits_engine
+
+template <unsigned long long _Xp, size_t _Rp>
+struct __log2_imp
+{
+    static const size_t value = _Xp & ((unsigned long long)(1) << _Rp) ? _Rp
+                                           : __log2_imp<_Xp, _Rp - 1>::value;
+};
+
+template <unsigned long long _Xp>
+struct __log2_imp<_Xp, 0>
+{
+    static const size_t value = 0;
+};
+
+template <size_t _Rp>
+struct __log2_imp<0, _Rp>
+{
+    static const size_t value = _Rp + 1;
+};
+
+template <class _UIntType, _UIntType _Xp>
+struct __log2
+{
+    static const size_t value = __log2_imp<_Xp,
+                                         sizeof(_UIntType) * __CHAR_BIT__ - 1>::value;
```

Comment:
doesn't build for me on Windows: `__CHAR_BIT__` is not standard

Reply:
Ah, missed that. There's a compat header in libc++ that defines it to `CHAR_BIT` (no underscores) which is in `<limits.h>`. Fixed by copying that hunk in.

Reply:
still doesn't build for me. Looks like the fix broke intrinsics

Reply:
Fixed on windows.